Updating team workflow

choldgraf opened the pull request:
We have been working on a team workflow to track our high/low-level issues, and to coordinate our work more effectively (and asynchronously).

The goal of this workflow is to be lightweight, but provide enough structure that we have a bit more "signal to noise" in order to figure out what we should work on next (and what others are working on).

I'd love feedback from @2i2c-org/tech-team about this! It doesn't have to be perfect, but good enough for us to try out over the next few weeks.

related to https://github.com/2i2c-org/meta/issues/165

damianavila commented on practices/coordination.md:
```diff
@@ -12,48 +12,86 @@ Here are the goals we optimize for in organizing our team coordination practices
+Here's a quick summary:
 
-A new project should be created for anything with a specific scope and set of deliverables that spans a reasonably long amount of time. 2i2c projects should all feed into the [2i2c organizational strategy and goals](../about/strategy.md).
+* [Deliverables](coordination:active-backlog) are discrete units of value that we bring to others. To be ready for work, they should be scoped properly, with enough information to complete them (e.g., a set of [tasks](coordination:tasks)).
```
I think the link (coordination:active-backlog) is incorrect here...

choldgraf replied:
Ah so in MyST Markdown, when being used with Sphinx, the markdown links behave like `:any:` roles in Sphinx. So this means that you can include either:

- A relative path to another file (e.g., `../index.md`)
- A URL (e.g. `https://google.com`)
- A cross-reference (e.g., `mylabel`)

In this case, I'm referencing the `coordination:active-backlog` section which is defined here:

https://github.com/2i2c-org/team-compass/pull/73/files#diff-b0c9a716af7c8f8c76b0af475a8b92ce614b96b340678951c855a7febad7654dR62

damianavila replied:
Yep, I quickly noted you wanted to cross-reference some other section... but my comment was triggered because I think the `coordination:active-backlog` reference is incorrect. Don't you want to reference the  `coordination: deliverables` section?

choldgraf replied:
ahhhh yep you are correct! I looked for a typo but hadn't considered that I'd just totally referenced the wrong section :-) thanks for spotting

damianavila replied:
I think you still did not fix this one in your latest commits, right?

choldgraf replied:
weird - you are correct! I just committed an update now

damianavila commented on practices/coordination.md:
```diff
@@ -12,48 +12,86 @@ Here are the goals we optimize for in organizing our team coordination practices
-For example, "As a 2i2c Hub User, I want to install a package on my own."
-Deliverables should be accomplishable in one or two sprints, and may have many tasks associated with them.
+In general, there should be only two or three deliverables per team member on this board.
+It should not become so heavily-populated that it is hard to keep track of deliverables!
```
You actually built a dev project board and an org project board... I guess you only want to mention the dev one here because of the dev-focused nature of this part of doc, am I right?

choldgraf replied:
yep correct - maybe we should add a "note" about the org one as well though. I am starting off just describing the "dev workflow" but we should generalize it over time.

choldgraf replied:
I've generalized this slightly and noted the locations of both project backlogs in the latest commit

damianavila commented on practices/coordination.md:
```diff
@@ -12,48 +12,94 @@ Here are the goals we optimize for in organizing our team coordination practices
-Tasks are verbs, not outcomes, and they should be extremely concrete and obvious what needs to be done.
-In addition, they should be accomplishable in one setting.
+- The [2i2c Development Projects Backlog](https://github.com/orgs/2i2c-org/projects/7) contains deliverables for our development-focused projects.
+- The [Organizatipnal Project Backlog](https://github.com/2i2c-org/meta/projects/6?fullscreen=true) contains deliverables that are organization-wide and not related to development.
```
Maybe remove the full-screen from this last link? Or actually, add that to the previous link. The idea is to stay consistent.

damianavila replied:
@choldgraf I still see the discrepancy here, even after your :+1:

choldgraf commented:
I've made a couple of changes:

- Addressed the comments that @damianavila brought up above
- Renamed the `Active Development Backlog` to instead be called the `Activity Backlog`. I feel like this is shorter and more memorable. Do others agree?

choldgraf commented on practices/coordination.md:
```diff
@@ -12,48 +12,94 @@ Here are the goals we optimize for in organizing our team coordination practices
+
+- {guilabel}`Needs Discussion/Refinement`: Deliverables that are high-priority but un-refined. Our goal should be having discussion and doing research in order to get these deliverables ready for work.
+- {guilabel}`Ready to Work`: Deliverables that are well-scoped and have a clear path forward, and are thus ready to implement. As deliverables in {guilabel}`In progress` are completed, we should replace them with deliverables from this column. Generally speaking, deliverables near the top have higher priority than those at the bottom.
+- {guilabel}`In progress`: Deliverables that we are currently working towards. This means that we should be generating [Tasks](coordination:tasks) in our [Active Development Board](https://github.com/orgs/2i2c-org/projects/5) to complete this deliverable (more on this below).
```
```suggestion
- {guilabel}`In progress`: Deliverables that we are currently working towards. This means that we should be generating [Tasks](coordination:tasks) in our [Activity Backlog](https://github.com/orgs/2i2c-org/projects/5) to complete this deliverable (more on this below).
```

choldgraf commented on practices/coordination.md:
```diff
@@ -12,48 +12,94 @@ Here are the goals we optimize for in organizing our team coordination practices
+Here's a quick summary:
 
-A new project should be created for anything with a specific scope and set of deliverables that spans a reasonably long amount of time. 2i2c projects should all feed into the [2i2c organizational strategy and goals](../about/strategy.md).
+* [Deliverables](coordination:active-backlog) are discrete units of value that we bring to others. To be ready for work, they should be scoped properly, with enough information to complete them (e.g., a set of [tasks](coordination:tasks)).
```
```suggestion
* [Deliverables](coordination:deliverables) are discrete units of value that we bring to others. To be ready for work, they should be scoped properly, with enough information to complete them (e.g., a set of [tasks](coordination:tasks)).
```

choldgraf commented:
OK I think this language is in pretty good shape for an "alpha level" process. I requested reviews from @yuvipanda / @GeorgianaElena / @consideRatio to see if this sounds like a reasonable process to try! Since we'll be integrating deliverables from everybody's projects in one space, I want to make sure this sounds like a sensible approach for the team.

Here's a place to preview the plan:

https://2i2c-team-compass--73.org.readthedocs.build/en/73/practices/coordination.html#team-development-workflow

consideRatio commented on practices/coordination.md:
```diff
@@ -12,48 +12,94 @@ Here are the goals we optimize for in organizing our team coordination practices
-Tasks are verbs, not outcomes, and they should be extremely concrete and obvious what needs to be done.
-In addition, they should be accomplishable in one setting.
+- The [2i2c Development Projects Backlog](https://github.com/orgs/2i2c-org/projects/7) contains deliverables for our development-focused projects.
+- The [Organizatipnal Project Backlog](https://github.com/2i2c-org/meta/projects/6?fullscreen=true) contains deliverables that are organization-wide and not related to development.
```
```suggestion
- The [Organizational Project Backlog](https://github.com/2i2c-org/meta/projects/6?fullscreen=true) contains deliverables that are organization-wide and not related to development.
```

consideRatio reviewed (commented):
I'm very happy about this process and how explicit it is!! I like it!

I have a platform specific question about how we couple and couple deliverables and tasks to the backlog boards. My mental model is that we would...

1. Let deliverables and task (github issues) have the label "deliverable" and "task", and they can exist across the org in various projects.
2. We have automation in the github project boards adding new "deliverable" and "task" labelled issues across the org to the respective github project board into the needs discussion column (project backlog / activity backlog).
3. We then manually move the backlog entries around between columns, but we have automation to automatically move github issues being closed to the done column.
4. An fancy option to step 3 would be to let movement of between columns associate with adding/removing labels representing the column as well. Then automation need to be two way so that movement in the project board update the issue, and changes in the issue move the issue in the project board. A benefit of this would be the visibility of the status an issue has in the associated project board by just inspecting the issue itself.

choldgraf commented on practices/coordination.md:
```diff
@@ -12,48 +12,94 @@ Here are the goals we optimize for in organizing our team coordination practices
+
+The Activity Backlog is broken down into these columns:
+
+- {guilabel}`To Do` Tasks that are ready to be worked on. Roughly speaking, tasks higher on the list are of higher priority.
+- {guilabel}`Needs Discussion` Tasks that require some team discussion around _implementation_. If there are higher-level discussions about the deliverable itself, perhaps the deliverable is not yet ready to be worked on.
+- {guilabel}`In progress` A task that a team member is currently working towards. When you move a task to In progress, indicate that you are working on it (either by adding your username to the card, or “assigning yourself” in the issue/PR associated with it.
+- {guilabel}`Needs review` Tasks that have an implementation that require feedback from others. For example, reviews for pull requests.
+- {guilabel}`Done` Tasks that are complete! When you move a task here, make sure to update any relevant deliverables.
```
huh weird - it is supposed to render like with a little "box" - maybe I need to release a new version of the sphinx book theme haha. I will try that

choldgraf commented on practices/coordination.md:
```diff
@@ -12,48 +12,94 @@ Here are the goals we optimize for in organizing our team coordination practices
-Team goals are high-level goals (AKA, there's no single action that will "complete" them) and span a longer period of time.
-Goals generally contain pointers to other issues / PRs / discussions where the work of more specific tasks is getting done.
-You should regularly communicate with other team members about which goals you're focusing on (e.g., via Slack, GitHub Issues, or via the [](coordination:team-syncs))
+A deliverable is ready to work (and can thus be added to the Project Backlog) when it has the following properties (adapted from the [INVEST methodology](https://agileforall.com/new-to-agile-invest-in-good-user-stories/)).
```
Ah good point. My one concern is that if we automatically add anything w/ `Deliverable` on it, then it'll become extremely long (depending on how many deliverables we have across all of our repositories). What do you think about this?

consideRatio replied:
Hmmm I'm a bit confused - are you commenting regarding automation associated with labels or the suggested change about "added" vs "advanced to that column"?

If we don't make this update, I'm considering how the column in the backlog about discussion is meant to be used. First add ready for work, and then it can transition back to needs discussion/refinement?

consideRatio replied:
@choldgraf regarding automation of labels, I'm not sure. If we don't, it means we don't really keep track of our deliverables in one place as i imagined the project backlog would do.

That would be fine, but it raises a question about what deliverables and when to add/remove them from the project board.

choldgraf replied:
I think I had imagined something like:

|               | Need more info                     | Ready for Work       |
|---------------|------------------------------------|----------------------|
| Low Priority  | Stays in issues                    | Put in Ready to work |
| High Priority | Put in Needs Discussion/Refinement | Put in Ready to work |

However, I think you are suggesting a simpler model, like:

|               | Need more info                     | Ready for Work       |
|---------------|------------------------------------|----------------------|
| Low Priority  | Put in Needs Discussion/Refinement | Put in Ready to work |
| High Priority | Put in Needs Discussion/Refinement | Put in Ready to work |

Is that right? If so, I think I'm 👍 on trying your approach out, and seeing how it goes. We can use GitHub labels to tag priority high/low. What do you think?

consideRatio replied:
Beautiful overview! Yes, that is what I meant and I now have a clear grasp on what I didn't before about what you meant!

I think both options make sense as long as they were a bit explicit, like both options now are.

I've just assumed it would be plausible to setup such automation, but I haven't done it. I think we should go for the non-automated approach until its proven viable. Looking into things I found some indication of viability but not by GitHub's official setup, but through https://github.com/philschatz/project-bot.

choldgraf replied:
Quick point on automation: I think the automation would be quite easy using ghapi (https://ghapi.fast.ai/) for example.

choldgraf commented on practices/coordination.md:
```diff
@@ -12,48 +12,94 @@ Here are the goals we optimize for in organizing our team coordination practices
-Goals generally contain pointers to other issues / PRs / discussions where the work of more specific tasks is getting done.
-You should regularly communicate with other team members about which goals you're focusing on (e.g., via Slack, GitHub Issues, or via the [](coordination:team-syncs))
+A deliverable is ready to work (and can thus be added to the Project Backlog) when it has the following properties (adapted from the [INVEST methodology](https://agileforall.com/new-to-agile-invest-in-good-user-stories/)).
 
-:::{tip}
-When opening a new issue in the 2i2c `pilot-hubs/` repository, first check if it would be better to add it to the to-do list of a pre-existing goal.
-:::
+- **Have a short title and description:** Deliverables should be glanceable and have enough context that the reader can quickly understand the scope.
+- **Have one or more [user stories](https://www.atlassian.com/agile/project-management/user-stories):** User stories should define who benefits from the deliverable, and why they want it.
+- **Have completion criteria:** We have clear completion criteria for this deliverable to denote it as “done”.
+- **Have tasks to complete it**: Deliverables should have a set of tasks, which are actions needed to complete the deliverable.
+
+#### How refinement happens
+
+_Deliverable Refinement_ is the process of improving the scoping, context, and structure of our Deliverables issues so that they are ready for us to work on them. When a deliverable is created, it may not have all of the information needed to take the next step. Adding that information is the goal of Deliverable Refinement.
+
+All team members are expected to participate in issue refinement, though the more experience and higher-level you are, the more you should be contributing to this process.
```
good point - I'll call it "deliverable refinement" since I think that is a more precise term

choldgraf commented on practices/coordination.md:
```diff
@@ -12,48 +12,94 @@ Here are the goals we optimize for in organizing our team coordination practices
+The Activity Backlog is broken down into these columns:
+
+- {guilabel}`To Do` Tasks that are ready to be worked on. Roughly speaking, tasks higher on the list are of higher priority.
+- {guilabel}`Needs Discussion` Tasks that require some team discussion around _implementation_. If there are higher-level discussions about the deliverable itself, perhaps the deliverable is not yet ready to be worked on.
```
good call - I will rename the column!

choldgraf commented:
@consideRatio most comments addressed! (and I released a new sphinx book theme, which fixed the guilabel thing)

I think the main outstanding question is whether to _automatically_ add deliverables to the project backlog, or whether we should do so manually. [See here for conversation on this](https://github.com/2i2c-org/team-compass/pull/73#discussion_r618458848). Would also be curious to hear what @damianavila thinks about it

damianavila commented:
> I think the main outstanding question is whether to automatically add deliverables to the project backlog, or whether we should do so manually. See here for conversation on this. Would also be curious to hear what @damianavila thinks about it

This is a good question. I really like automation... BUT, I think we should try this manually first for a few weeks to see how that works. I am also afraid of too many deliverables on the first column that could confuse people, so let's start with a few high-prio deliverables at the beginning that we add manually. Then, if everything flows, and when people are used to the process, we can put ALL the deliverables in the first column and automate the whole thing as @consideRatio proposed.
I really like the idea of using labels to 1) establish prioritization across deliverables, 2) automating the movement between columns, and 3) quickly check upstream information (as an issue, I see which deliverable I belong to).

choldgraf commented:
OK - in the latest commit I've added Table 1 from my comment above. This is the one that suggests we manually put high-priority issues that need refinement in the respective column (rather than automating it). I agree that over time we should revisit this and decide if we want to tweak it a bit.

I think that this PR is ready to go from my end, unless others have tweaks they'd like to see. I think the question is "is this good enough for our team to give it a try?"

In the future, we should discuss more the options of automating, using priority labels, etc.

damianavila reviewed (approved):
I think this is ready for a first try!

yuvipanda commented on practices/coordination.md:
```diff
@@ -12,48 +12,109 @@ Here are the goals we optimize for in organizing our team coordination practices
-Team goals are high-level goals (AKA, there's no single action that will "complete" them) and span a longer period of time.
-Goals generally contain pointers to other issues / PRs / discussions where the work of more specific tasks is getting done.
-You should regularly communicate with other team members about which goals you're focusing on (e.g., via Slack, GitHub Issues, or via the [](coordination:team-syncs))
+All team members are expected to participate in deliverable refinement, though the more experience and higher-level you are, the more you should be contributing to this process.
```
```suggestion
All team members are expected to participate in deliverable refinement, though the more experienced and higher-level you are, the more you should be contributing to this process.
```

yuvipanda reviewed (approved):
I appreciate the level of detail in here - just enough without being overwhelming. +1 to trying this for a while and seeing how this goes.

GeorgianaElena reviewed (approved):
Thanks for putting this worflow together and for describing it so well. I like it! :+1: 
Let's see it in action :rocket:

choldgraf commented:
thanks all for this helpful feedback and all of your advice! Let's merge this in, and start giving it a shot as a team :-) 

I will update some of our team processes / docs / infrastructure today and then try to close out this deliverable :-)
